src/2020-06-08-mixing-modules-and-export.Rmd

```diff
@@ -0,0 +1,106 @@
+---
+title: "Mixing Rcpp modules and Rcpp attributes"
+author: "Bob Jansen"
+license: GPL (>= 2)
+tags: basics attributes modules
+summary: How to use Rcpp modules in combination with Rcpp attributes
+---
+
+### Introduction
+
+With Rcpp modules (Eddelbuettel and Francois, 2017) it is easy to expose
+\proglang{C++} classes and functions to \proglang{R}. This note describes how
+to use exported classes in conjunction with functions exported using Rcpp
+attributes (Eddelbuettel and Francois, 2017) using Rcpp attributes
+(Eddelbuettel and Francois, 2017) through the use of
+\code{RCPP_EXPOSED\*} macros.
+
+Exporting \proglang{C++} functions is simple thanks to Rcpp attributes. Just
+add a single specially formatted comment line to each function as follows:
+
+```{r chunk-1, engine="Rcpp"}
+#include <Rcpp.h>
+
+// [[Rcpp::export]]
+void shout(std::string message) {
+  Rcpp::Rcout << message << "!" << std::endl;
+}
+```
+Now, calling this function from \code{R} is as easy as one can hope:
+```{r}
+shout("Hello World")
+```
+
+Exporting \proglang{C++} classes can be done using Rcpp modules. It uses a
+different mechanism and requires a bit more work. The simple class \code{Echo}
+below has a \code{get()} method which returns the original constructor
+parameter.
+
+```{r chunk-2, engine="Rcpp", eval=FALSE}
+#include <Rcpp.h>
+#include <string>
+
+using namespace Rcpp;
+
+class Echo {
+private:
+    std::string message;
+public:
+    Echo(std::string message) : message(message) {}
+
+    std::string get() {
+      return message;
+    }
+};
+```
+
+This simple class can now be exposed to R by specifing the constructors and the
+methods that should be callable from R along
+
+```{r chunk-3, engine="Rcpp", eval=FALSE}
+RCPP_MODULE(echo_module) {
+    class_<Echo>("Echo")
+    .constructor<std::string>()
+    .method("get", &Echo::get)
+    ;
+};
+```
+
+Unfortunately, combining these two snippets as is creates a problem. The Rcpp
+attributes machinery that exports \code{shout()} will not be automagically
+aware of the \code{Echo} class. This will cause a _link time error_ when
```

Comment:
Isn't it a compile error?  The compiler has as `as<>` and `wrap()` template to fall back to.  Defensively you could call it a build-time error :)

Reply:
I made a more precise description of the behaviour I witnessed.